{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c33e6510-3a10-4bf8-8724-b1ab7fa08cbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "import requests\n",
    "from time import sleep\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "165bff00-2b93-4a24-9080-2665f7c3178b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rsid to process: 5\n"
     ]
    }
   ],
   "source": [
    "# get list of rsid-s for scraping from the file\n",
    "with open('data/list-test.txt') as fh:\n",
    "    ls_rsid = [rsid.strip() for rsid in fh.readlines()]\n",
    "    \n",
    "print('Number of rsid to process:', len(ls_rsid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d4606713-8b64-4a5d-9d3f-c7bcf12c9b6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b2152246-c8c3-453b-8fc5-395294c6ce9d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  #   rsid         ref_allele    alt_alleles\n",
      "  0   rs5128        G             C\n",
      "  1   rs7412        C             T\n",
      "WARNING: rsid = rs1799752: problem with splitting of allele string, length of pair is 94 instead of expected 3\n",
      "Alleles: ins(T)11GAGACGGAGTCTCGCTCTGTCG…\n",
      "  2   rs1799752     -             -\n",
      "  3   rs186838828   T             A\n",
      "  4   rs191667986   T             C\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>organism</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_genome</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_alleles</th>\n",
       "      <th>variation_type</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>gene_consequence</th>\n",
       "      <th>publications</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs5128</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>chr11:116832924</td>\n",
       "      <td>GRCh38.p13</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Not Reported in ClinVar</td>\n",
       "      <td>APOC3 : 3 Prime UTR Variant</td>\n",
       "      <td>47 citations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs7412</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>chr19:44908822</td>\n",
       "      <td>GRCh38.p13</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Reported in ClinVar</td>\n",
       "      <td>APOE : Missense Variant</td>\n",
       "      <td>386 citations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1799752</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>chr17:63488530-63488543</td>\n",
       "      <td>GRCh38.p13</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Indel</td>\n",
       "      <td>Reported in ClinVar</td>\n",
       "      <td>ACE : Intron Variant</td>\n",
       "      <td>74 citations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs186838828</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>chr2:166195246</td>\n",
       "      <td>GRCh38.p13</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Reported in ClinVar</td>\n",
       "      <td>SCN1A-AS1 : Intron Variant; SCN9A : 3 Prime UT...</td>\n",
       "      <td>0 citations</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs191667986</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>chr2:166195355</td>\n",
       "      <td>GRCh38.p13</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Reported in ClinVar</td>\n",
       "      <td>SCN1A-AS1 : Intron Variant; SCN9A : 3 Prime UT...</td>\n",
       "      <td>0 citations</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          rsid      organism                 position  ref_genome ref_allele  \\\n",
       "0       rs5128  Homo sapiens          chr11:116832924  GRCh38.p13          G   \n",
       "1       rs7412  Homo sapiens           chr19:44908822  GRCh38.p13          C   \n",
       "2    rs1799752  Homo sapiens  chr17:63488530-63488543  GRCh38.p13          -   \n",
       "3  rs186838828  Homo sapiens           chr2:166195246  GRCh38.p13          T   \n",
       "4  rs191667986  Homo sapiens           chr2:166195355  GRCh38.p13          T   \n",
       "\n",
       "  alt_alleles variation_type    clinical_significance  \\\n",
       "0           C            SNV  Not Reported in ClinVar   \n",
       "1           T            SNV      Reported in ClinVar   \n",
       "2           -          Indel      Reported in ClinVar   \n",
       "3           A            SNV      Reported in ClinVar   \n",
       "4           C            SNV      Reported in ClinVar   \n",
       "\n",
       "                                    gene_consequence   publications  \n",
       "0                        APOC3 : 3 Prime UTR Variant   47 citations  \n",
       "1                            APOE : Missense Variant  386 citations  \n",
       "2                               ACE : Intron Variant   74 citations  \n",
       "3  SCN1A-AS1 : Intron Variant; SCN9A : 3 Prime UT...    0 citations  \n",
       "4  SCN1A-AS1 : Intron Variant; SCN9A : 3 Prime UT...    0 citations  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def handle_rsid(rsid):\n",
    "    \"\"\"\n",
    "    The function gets 1 rsid (as string or integer), mades query in usual way to NCBI server,\n",
    "    extracts some data from the code and packs them to Pandas Series, that is returned from the function.\n",
    "    The functin is not ideal in terms of defence against server error, and sometimes can break down.\n",
    "    But if server response is good, it can process a thousand of rsid-s.    \n",
    "    \"\"\"\n",
    "    \n",
    "    # get html-code of the page\n",
    "    url = 'https://www.ncbi.nlm.nih.gov/snp/' + str(rsid)\n",
    "    \n",
    "    # Made response to the server. If there is a problem, try again.\n",
    "    response = requests.get(url)\n",
    "    while response.status_code != 200:\n",
    "        print(f'WARNING: rsid = {rsid}, response_code = {response.status_code}')\n",
    "        sleep(3)\n",
    "        response = requests.get(url)\n",
    "    \n",
    "    # parse response-object to BeautifulSoup-object\n",
    "    data = response.text\n",
    "    soup = BeautifulSoup(data, 'html.parser')\n",
    "    \n",
    "    # preliminaty extract from html-code part of code with necessary values by specified tags and their attributes\n",
    "    ls_dl = soup.find('div', attrs={'class': 'summary-box usa-grid-full'}).find_all('dl', attrs={'class': 'usa-width-one-half'})\n",
    "\n",
    "    # extract some values and place them into variable (left half of data block on the page)\n",
    "    ls_dt = ls_dl[0].find_all('dt')\n",
    "    ls_dd = ls_dl[0].find_all('dd')\n",
    "    for el_dt, el_dd in zip(ls_dt, ls_dd):\n",
    "        el_dt = el_dt.text.strip()\n",
    "        if el_dt == 'Organism':\n",
    "            organism = el_dd.text.strip()\n",
    "        elif el_dt == 'Position':\n",
    "            ls_span = el_dd.find_all('span')\n",
    "            pos = ls_span[0].text.strip()\n",
    "            ref_genome = ls_span[1].text.strip(\" ( )\")\n",
    "        elif el_dt == 'Alleles':\n",
    "            alleles = el_dd.text.strip()\n",
    "            ls_allele_pairs = alleles.split(' / ')\n",
    "            ref_allele = ''\n",
    "            alt_alleles = ''\n",
    "            for pair in ls_allele_pairs:\n",
    "                if len(pair) != 3:\n",
    "                    print(f\"WARNING: rsid = {rsid}: problem with splitting of allele string, length of pair is {len(pair)} instead of expected 3\")\n",
    "                    print(f\"Alleles: {pair[:30]}{'…' if len(pair)>30 else ''}\")\n",
    "                    ref_allele = '-'\n",
    "                    alt_alleles = '-'\n",
    "                else:\n",
    "                    try:\n",
    "                        al_1, al_2 = pair.split('>')\n",
    "                    except ValueError as ex:\n",
    "                        print(f'CAUGHT ERROR: rsid = {rsid}: {ex}')\n",
    "                        ref_allele = '-'\n",
    "                        alt_alleles = '-'\n",
    "                        break\n",
    "                    else:\n",
    "                        if ref_allele == '':\n",
    "                            ref_allele = al_1\n",
    "                        elif al_1 != ref_allele:\n",
    "                            ref_allele += ', ' + al_1\n",
    "                            print(f'For rsid = {rsid}, there are more than one reference allele')\n",
    "                            # raise ValueError(f'For rsid = {rsid}, there is more than one reference allele')\n",
    "\n",
    "                        if alt_alleles == '':\n",
    "                            alt_alleles = al_2\n",
    "                        else:\n",
    "                            alt_alleles += ', ' + al_2\n",
    "\n",
    "        elif el_dt == 'Variation Type':\n",
    "            variation_type = el_dd.text.lstrip().split('\\n')[0]\n",
    "    \n",
    "    # extract some values and place them into variable (right half of data block on the page)\n",
    "    ls_dt = ls_dl[1].find_all('dt')\n",
    "    ls_dd = ls_dl[1].find_all('dd')\n",
    "    for el_dt, el_dd in zip(ls_dt, ls_dd):\n",
    "        el_dt = el_dt.text.strip()\n",
    "        if el_dt == 'Clinical Significance':\n",
    "            clinical_significance = el_dd.text.strip()\n",
    "        elif el_dt == 'Gene : Consequence':\n",
    "            gene_consequence = '; '.join(map(lambda tag:tag.text.strip(), el_dd.findAll() ) )    #el_dd.text.strip()\n",
    "        elif el_dt == 'Publications':\n",
    "            publications = ' '.join(st.strip() for st in el_dd.text.strip().split('\\n'))\n",
    "\n",
    "    # pack variables into Pandas series\n",
    "    ser = pd.Series({\n",
    "        'rsid': rsid,\n",
    "        'organism': organism,\n",
    "        'position': pos,\n",
    "        'ref_genome': ref_genome,\n",
    "        'ref_allele': ref_allele,\n",
    "        'alt_alleles': alt_alleles,\n",
    "        'variation_type': variation_type,\n",
    "        'clinical_significance': clinical_significance,\n",
    "        'gene_consequence': gene_consequence,\n",
    "        'publications': publications\n",
    "    })\n",
    "    \n",
    "    return ser\n",
    "\n",
    "\n",
    "# print header for preview inside this Jupyter notebook\n",
    "print(\"  #   rsid         ref_allele    alt_alleles\")\n",
    "\n",
    "start_processing = 0\n",
    "for i, rsid in enumerate(ls_rsid[start_processing:], start=start_processing):\n",
    "    \n",
    "    # optimization so that we don't send request with the same rsid to NCBI several times\n",
    "    if not df.empty and (rsid in df.rsid.to_list()):\n",
    "        ser = df[df.rsid == rsid].iloc[0]\n",
    "    else:\n",
    "        ser = handle_rsid(rsid)\n",
    "        \n",
    "    ser.name = i\n",
    "    df = df.append(ser)\n",
    "    \n",
    "    # preview\n",
    "    print(f\"{i:3}   {rsid:11}   {ser.ref_allele}             {ser.alt_alleles}\")\n",
    "    \n",
    "    sleep(3)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "34e4f553-c9c4-4b6a-8c89-861666be92b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.index.name='#'\n",
    "df.to_csv(\"output/data_NCBI.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
