{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2d7734d4-dce1-418d-a66d-5f647a3b9687",
   "metadata": {},
   "source": [
    "Code of notebook [NCBI-scraper-simple](https://github.com/Lady3mlnm/NCBI-scraper-simple/blob/main/NCBI-scraping.ipynb)<br />\n",
    "query to get variations for gene: [https://www.ncbi.nlm.nih.gov/clinvar/](https://www.ncbi.nlm.nih.gov/clinvar/), insert \"insensitivity to pain SCN9A\", download as \"UI List\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7a2c5c70-52bf-446e-8efd-979893ec7959",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "from time import sleep\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7b444872-dcdf-4cfd-aad3-8438721bea38",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.display.max_colwidth = 60"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "421a6556-c135-4159-ace8-efd7fda3e261",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variations to process: 4\n",
      "['530654', '331895', '471143', '892607']\n"
     ]
    }
   ],
   "source": [
    "# read from file list of vatiations\n",
    "with open('input/UI List -test.txt', 'r') as fh:\n",
    "    ls_variations = [el.strip() for el in fh.readlines()]\n",
    "    \n",
    "print('Number of variations to process:', len(ls_variations))\n",
    "print(ls_variations[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbde05b0-f9f2-4364-b4ec-774cb2622a5d",
   "metadata": {},
   "source": [
    "<br />\n",
    "<br />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "510f9125-7c97-4990-8294-08ac982fc87d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def handle_rsid(rsid, url_rsid, variation, url_variation):\n",
    "    \"\"\"\n",
    "    The function gets 1 rsid (as string or integer), mades query in usual way to NCBI server,\n",
    "    extracts some data from the code and packs them to Pandas Series, that is returned from the function.\n",
    "    The functin is not ideal in terms of defence against server error, and sometimes can break down.\n",
    "    But if server response is good, it can process a thousand of rsid-s.    \n",
    "    \"\"\"\n",
    "    \n",
    "    # get html-code of the page\n",
    "    # url = 'https://www.ncbi.nlm.nih.gov/snp/' + str(rsid)\n",
    "    \n",
    "    # Made response to the server. If there is a problem, try again.\n",
    "    response = requests.get(url_rsid)\n",
    "    while response.status_code != 200:\n",
    "        print(f'WARNING: rsid = {rsid}, response_code = {response.status_code}')\n",
    "        sleep(3)\n",
    "        response = requests.get(url)\n",
    "    \n",
    "    # parse response-object to BeautifulSoup-object\n",
    "    data = response.text\n",
    "    soup = BeautifulSoup(data, 'html.parser')\n",
    "    \n",
    "    # preliminaty extract from html-code part of code with necessary values by specified tags and their attributes\n",
    "    ls_dl = soup.find('div', attrs={'class': 'summary-box usa-grid-full'}).find_all('dl', attrs={'class': 'usa-width-one-half'})\n",
    "\n",
    "    # extract some values and place them into variable (left half of data block on the page)\n",
    "    ls_dt = ls_dl[0].find_all('dt')\n",
    "    ls_dd = ls_dl[0].find_all('dd')\n",
    "    for el_dt, el_dd in zip(ls_dt, ls_dd):\n",
    "        el_dt = el_dt.text.strip()\n",
    "        if el_dt == 'Organism':\n",
    "            organism = el_dd.text.strip()\n",
    "        elif el_dt == 'Position':\n",
    "            ls_span = el_dd.find_all('span')\n",
    "            pos = ls_span[0].text.strip()\n",
    "            ref_genome = ls_span[1].text.strip(\" ( )\")\n",
    "        elif el_dt == 'Alleles':\n",
    "            alleles = el_dd.text.strip()\n",
    "            ls_allele_pairs = alleles.split(' / ')\n",
    "            ref_allele = ''\n",
    "            alt_alleles = ''\n",
    "            for pair in ls_allele_pairs:\n",
    "                if len(pair) != 3:\n",
    "                    print(f\"WARNING: rsid = {rsid}: problem with splitting of allele string, length of pair is {len(pair)} instead of expected 3\")\n",
    "                    print(f\"Alleles: {pair[:30]}{'…' if len(pair)>30 else ''}\")\n",
    "                    ref_allele = '-'\n",
    "                    alt_alleles = '-'\n",
    "                else:\n",
    "                    try:\n",
    "                        al_1, al_2 = pair.split('>')\n",
    "                    except ValueError as ex:\n",
    "                        print(f'CAUGHT ERROR: rsid = {rsid}: {ex}')\n",
    "                        ref_allele = '-'\n",
    "                        alt_alleles = '-'\n",
    "                        break\n",
    "                    else:\n",
    "                        if ref_allele == '':\n",
    "                            ref_allele = al_1\n",
    "                        elif al_1 != ref_allele:\n",
    "                            ref_allele += ', ' + al_1\n",
    "                            print(f'For rsid = {rsid}, there are more than one reference allele')\n",
    "                            # raise ValueError(f'For rsid = {rsid}, there is more than one reference allele')\n",
    "\n",
    "                        if alt_alleles == '':\n",
    "                            alt_alleles = al_2\n",
    "                        else:\n",
    "                            alt_alleles += ', ' + al_2\n",
    "\n",
    "        elif el_dt == 'Variation Type':\n",
    "            variation_type = el_dd.text.lstrip().split('\\n')[0]\n",
    "    \n",
    "    # extract some values and place them into variable (right half of data block on the page)\n",
    "    ls_dt = ls_dl[1].find_all('dt')\n",
    "    ls_dd = ls_dl[1].find_all('dd')\n",
    "    for el_dt, el_dd in zip(ls_dt, ls_dd):\n",
    "        el_dt = el_dt.text.strip()\n",
    "        if el_dt == 'Clinical Significance':\n",
    "            clinical_significance = el_dd.text.strip()\n",
    "        elif el_dt == 'Gene : Consequence':\n",
    "            gene_consequence = '; '.join(map(lambda tag:tag.text.strip(), el_dd.findAll() ) )    #el_dd.text.strip()\n",
    "        elif el_dt == 'Publications':\n",
    "            publications = ' '.join(st.strip() for st in el_dd.text.strip().split('\\n'))\n",
    "\n",
    "    # pack variables into Pandas series\n",
    "    ser = pd.Series({\n",
    "        'rsid': rsid,\n",
    "        'organism': organism,\n",
    "        'position': pos[3:] if pos.startswith('chr') else pos,\n",
    "        'ref_genome': ref_genome,\n",
    "        'ref_allele': ref_allele,\n",
    "        'alt_alleles': alt_alleles,\n",
    "        'variation_type': variation_type,\n",
    "        'clinical_significance': clinical_significance,\n",
    "        'gene_consequence': gene_consequence,\n",
    "        'publications': publications,\n",
    "        'url_rsid': url_rsid,\n",
    "        'variation': variation,\n",
    "        'url_variation': url_variation\n",
    "    })\n",
    "    \n",
    "    return ser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "05d51049-6aef-4a69-8df5-f904d1e68f81",
   "metadata": {},
   "outputs": [],
   "source": [
    "def handle_variation(variation:str):\n",
    "    url_variation = 'https://www.ncbi.nlm.nih.gov/clinvar/variation/' + variation + '/'\n",
    "    \n",
    "    # Made response to the server. If there is a problem, try again.\n",
    "    response = requests.get(url_variation)\n",
    "    while response.status_code != 200:\n",
    "        print(f'WARNING: rsid = {rsid}, response_code = {response.status_code}')\n",
    "        sleep(3)\n",
    "        response = requests.get(url)\n",
    "\n",
    "    # parse response-object to BeautifulSoup-object\n",
    "    data = response.text\n",
    "    soup = BeautifulSoup(data, 'html.parser')\n",
    "    \n",
    "    # extract element that is starting point for further searching\n",
    "    el = soup.find('dt', attrs={'class': 'hashelp'})\n",
    "    \n",
    "    # explore next elements to find rsid, if it present in the code\n",
    "    while True:\n",
    "        el = el.find_next_sibling()\n",
    "        if not el:\n",
    "            rsid = None\n",
    "            url_rsid = None\n",
    "            break\n",
    "        text = el.text.strip()\n",
    "        if text.startswith('dbSNP: rs'):\n",
    "            rsid = text[7:]\n",
    "            url_rsid = el.find('a')['href']\n",
    "            # print(url_rsid)\n",
    "            break\n",
    "\n",
    "    if rsid:\n",
    "        return handle_rsid(rsid, url_rsid, variation, url_variation)\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "# ser = handle_variation(ls_variations[1])\n",
    "# ser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a4de7dab-174a-4f18-b78e-dde479c2d4a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  #   variation rsid         ref_genome ref_allele   alt_alleles  gene_consequence\n",
      "  1     —\n",
      "  2   331895   rs186838828   GRCh38.p13     T            A        SCN1A-AS1 : Intron Variant; SCN9A : 3 Prime UTR Variant\n",
      "WARNING: rsid = rs1553473041: problem with splitting of allele string, length of pair is 4 instead of expected 3\n",
      "Alleles: delT\n",
      "  3   471143   rs1553473041  GRCh38.p13     -            -        SCN9A : Frameshift Variant; SCN1A-AS1 : Intron Variant\n",
      "  4   892607   rs372210358   GRCh38.p13     C            G        SCN9A : Synonymous Variant; SCN1A-AS1 : Intron Variant\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>organism</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_genome</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_alleles</th>\n",
       "      <th>variation_type</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>gene_consequence</th>\n",
       "      <th>publications</th>\n",
       "      <th>url_rsid</th>\n",
       "      <th>variation</th>\n",
       "      <th>url_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs186838828</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2:166195246</td>\n",
       "      <td>GRCh38.p13</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Reported in ClinVar</td>\n",
       "      <td>SCN1A-AS1 : Intron Variant; SCN9A : 3 Prime UTR Variant</td>\n",
       "      <td>0 citations</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/snp/rs186838828</td>\n",
       "      <td>331895</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/clinvar/variation/331895/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1553473041</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2:166199288</td>\n",
       "      <td>GRCh38.p13</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>Reported in ClinVar</td>\n",
       "      <td>SCN9A : Frameshift Variant; SCN1A-AS1 : Intron Variant</td>\n",
       "      <td>0 citations</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/snp/rs1553473041</td>\n",
       "      <td>471143</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/clinvar/variation/471143/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs372210358</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>2:166199392</td>\n",
       "      <td>GRCh38.p13</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>Reported in ClinVar</td>\n",
       "      <td>SCN9A : Synonymous Variant; SCN1A-AS1 : Intron Variant</td>\n",
       "      <td>0 citations</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/snp/rs372210358</td>\n",
       "      <td>892607</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/clinvar/variation/892607/</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           rsid      organism     position  ref_genome ref_allele alt_alleles  \\\n",
       "2   rs186838828  Homo sapiens  2:166195246  GRCh38.p13          T           A   \n",
       "3  rs1553473041  Homo sapiens  2:166199288  GRCh38.p13          -           -   \n",
       "4   rs372210358  Homo sapiens  2:166199392  GRCh38.p13          C           G   \n",
       "\n",
       "  variation_type clinical_significance  \\\n",
       "2            SNV   Reported in ClinVar   \n",
       "3       Deletion   Reported in ClinVar   \n",
       "4            SNV   Reported in ClinVar   \n",
       "\n",
       "                                          gene_consequence publications  \\\n",
       "2  SCN1A-AS1 : Intron Variant; SCN9A : 3 Prime UTR Variant  0 citations   \n",
       "3   SCN9A : Frameshift Variant; SCN1A-AS1 : Intron Variant  0 citations   \n",
       "4   SCN9A : Synonymous Variant; SCN1A-AS1 : Intron Variant  0 citations   \n",
       "\n",
       "                                        url_rsid variation  \\\n",
       "2   https://www.ncbi.nlm.nih.gov/snp/rs186838828    331895   \n",
       "3  https://www.ncbi.nlm.nih.gov/snp/rs1553473041    471143   \n",
       "4   https://www.ncbi.nlm.nih.gov/snp/rs372210358    892607   \n",
       "\n",
       "                                            url_variation  \n",
       "2  https://www.ncbi.nlm.nih.gov/clinvar/variation/331895/  \n",
       "3  https://www.ncbi.nlm.nih.gov/clinvar/variation/471143/  \n",
       "4  https://www.ncbi.nlm.nih.gov/clinvar/variation/892607/  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame()\n",
    "\n",
    "# print header for preview inside this Jupyter notebook\n",
    "print(\"  #   variation rsid         ref_genome ref_allele   alt_alleles  gene_consequence\")\n",
    "\n",
    "for i, variation in enumerate(ls_variations[:], start=1):\n",
    "    ser = handle_variation(variation)\n",
    "    if ser is not None:\n",
    "        print(f\"{i:3}   {variation:7}  {ser.rsid:12}  {ser.ref_genome:10}     {ser.ref_allele:3}          {ser.alt_alleles:7}  {ser.gene_consequence}\")\n",
    "        ser.name = i\n",
    "        df = df.append(ser)\n",
    "    else:\n",
    "        print(f\"{i:3}     —\")\n",
    "    \n",
    "    sleep(3)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "da61992f-93ba-41e7-8619-e48145fa0151",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.index.name='#'\n",
    "df.to_csv(\"output/info from variants.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c833c0c7-d4b8-4454-97d2-6f954f4028ff",
   "metadata": {},
   "source": [
    "<br />"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e97dc7fd-c61d-491d-8f57-3ad7812ca0ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "October 16, 14:23:14\n"
     ]
    }
   ],
   "source": [
    "# play beep to denote completion of the program\n",
    "import IPython.display as ipd\n",
    "import numpy as np\n",
    "\n",
    "# manually generated sound\n",
    "t = 1  # time is seconds\n",
    "beep = np.sin(2*np.pi*400*np.arange(10000*t)/10000)\n",
    "ipd.Audio(beep, rate=10000, autoplay=True)\n",
    "\n",
    "\n",
    "from datetime import datetime\n",
    "print(datetime.now().strftime(\"%B %d, %H:%M:%S\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
